File: apps/console/src/platform/domain/sides.ts
/**
 * Sides — the console's one side formatter (package 10b, v3 consolidated
 * plan). D17/D14/D15 land here: a `Side` is the discriminated model from the
 * match-card contract (§2.1) and the state-and-formatting contract (§6.1),
 * never a pre-joined string, and it is never parsed back OUT of one either
 * (`names.ts`'s split/rejoin round-trip is the anti-pattern this file
 * replaces for every site listed in package 10's scope).
 *
 * Two constructors meet the two engines where their wire already is:
 *  - `meetSideFromIds` — the meet engine already ships individual player ids
 *    + a roster, so a meet side gets full per-person fidelity here, with NO
 *    backend change (see D17 note in the package 10 report).
 *  - `sideFromWire` — the bracket engine's structured `SideDTO`
 *    (`shared/sides.py`, package 10a). Since v3 package 29 (V3-10-1) a
 *    bracket doubles pair arrives as TWO `PersonRef`s: the backend resolves
 *    `member_ids` against the bracket roster rather than shipping only the
 *    composite team name, so full per-partner fidelity now reaches this file
 *    from BOTH engines. A pair one member short arrives as
 *    `persons: [A], unresolved: pending_member` (V3-10-2), which is what
 *    `formatSideLines` turns into the known name plus "partner to be
 *    confirmed". Where a member id has no roster row the backend still sends
 *    the composite name as one person — the only honest label it has.
 *
 * `resolveFeederReference` is the seam back to the console's own identity
 * authority: the wire's `winner_of`/`loser_of` carries a RAW play-unit id
 * (state-and-formatting §6.3 — the backend does not duplicate the console's
 * round-label speller), and a caller with a `labelById` map (
 * `bracketLabels.ts`'s `buildPlayUnitLabels`) turns it into "Winner of QF1"
 * before formatting.
 *
 * `formatPersonName`/`formatSideName`/`sideNameLines` (retired from
 * `lib/names.ts`, V3-10-3) are a narrower compatibility shim for callers that
 * only ever held a pre-joined side string ("A / B") rather than the
 * structured wire `sides` this file otherwise expects — `MatchesSpreadsheet`,
 * `BracketPlayerFields`, `ParticipantPicker` and `RunCourtGrid` were outside
 * package 10c's scope (BracketMatchesTab, prior art for the full `Side`
 * conversion) and still work from joined strings. They preserve the operator-
 * entered name exactly (no reordering — names are not safely parseable from
 * whitespace) and stay presentation-only, same as their `names.ts` originals.
 */

/** One resolved (or dead) person reference. */
export interface PersonRef {
  id: string | null;
  name: string;
}

/** Why a side has no (or an incomplete) resolved person — match-card
 *  contract §2.1. `persons` and `unresolved` on `Side` are NOT mutually
 *  exclusive: `pending_member` carries both the known persons and this. */
export type UnresolvedSide =
  | { kind: 'bye' }
  | { kind: 'pending_member'; known: PersonRef[]; missing: number }
  | { kind: 'winner_of'; reference: string }
  | { kind: 'loser_of'; reference: string }
  | { kind: 'withheld' }
  | { kind: 'undetermined' };

export interface Side {
  persons: PersonRef[];
  unresolved: UnresolvedSide | null;
  seed: number | null;
  participantKey: string | null;
}

const UNDETERMINED: Side = {
  persons: [],
  unresolved: { kind: 'undetermined' },
  seed: null,
  participantKey: null,
};

/** Build a side from the meet engine's raw player ids + a roster name map.
 *  Every id resolves to its own `PersonLine` (one participant per line) —
 *  the meet wire already carries per-person identity, so this is a pure
 *  projection, never a join. */
export function meetSideFromIds(
  ids: readonly string[] | undefined,
  nameById: Record<string, string>,
): Side {
  if (!ids || ids.length === 0) return UNDETERMINED;
  return {
    persons: ids.map((id) => ({ id, name: nameById[id] ?? id })),
    unresolved: null,
    seed: null,
    participantKey: ids.length === 1 ? ids[0] : null,
  };
}

/** The minimal shape `sideFromWire` needs from `shared/sides.py`'s
 *  `SideDTO` — matched structurally so this file does not import the
 *  bracket-specific wire module (foundation layers stay dependency-light;
 *  callers pass their own `BracketDto.SideDTO`, which satisfies this shape). */
export interface WireSide {
  persons?: { id?: string | null; name: string }[] | null;
  unresolved?: {
    kind: UnresolvedSide['kind'];
    known?: { id?: string | null; name: string }[] | null;
    missing?: number | null;
    reference?: string | null;
  } | null;
  seed?: number | null;
  participantKey?: string | null;
}

/** Build a side from the operator wire's structured `SideDTO` (bracket,
 *  package 10a). Absent entirely on an older cached payload — callers fall
 *  back to `undefined` and should use the legacy `slot`/`side_a` reader. */
export function sideFromWire(dto: WireSide | null | undefined): Side {
  if (!dto) return UNDETERMINED;
  const persons = (dto.persons ?? []).map((p) => ({ id: p.id ?? null, name: p.name }));
  let unresolved: UnresolvedSide | null = null;
  const u = dto.unresolved;
  if (u) {
    switch (u.kind) {
      case 'bye':
        unresolved = { kind: 'bye' };
        break;
      case 'withheld':
        unresolved = { kind: 'withheld' };
        break;
      case 'undetermined':
        unresolved = { kind: 'undetermined' };
        break;
      case 'winner_of':
        unresolved = { kind: 'winner_of', reference: u.reference ?? '' };
        break;
      case 'loser_of':
        unresolved = { kind: 'loser_of', reference: u.reference ?? '' };
        break;
      case 'pending_member':
        unresolved = {
          kind: 'pending_member',
          known: (u.known ?? []).map((p) => ({ id: p.id ?? null, name: p.name })),
          missing: u.missing ?? 1,
        };
        break;
    }
  } else if (persons.length === 0) {
    unresolved = { kind: 'undetermined' };
  }
  return {
    persons,
    unresolved,
    seed: dto.seed ?? null,
    participantKey: dto.participantKey ?? null,
  };
}

/** Resolve a `winner_of`/`loser_of` side's raw play-unit-id reference to the
 *  console's own friendly identity spelling ("Winner of QF1"). A no-op for
 *  every other side shape. `labelById` is `bracketLabels.ts`'s
 *  `buildPlayUnitLabels()` map; an id with no entry passes through raw
 *  rather than throwing, matching the legacy `sideLabel` fallback. */
export function resolveFeederReference(
  side: Side,
  labelById: ReadonlyMap<string, string> | Record<string, string>,
): Side {
  if (side.unresolved?.kind !== 'winner_of' && side.unresolved?.kind !== 'loser_of') return side;
  const raw = side.unresolved.reference;
  const label = labelById instanceof Map ? labelById.get(raw) : (labelById as Record<string, string>)[raw];
  if (!label || label === raw) return side;
  return { ...side, unresolved: { ...side.unresolved, reference: label } };
}

/** One line per participant, in the fixed label wording (state-and-
 *  formatting §6.1 / match-card §2.1). Never `TBD`, `–`, `No players` or an
 *  empty string — an unresolved side always renders its label, in the same
 *  ink and size as a name (never a blank row). */
export function formatSideLines(side: Side): string[] {
  if (side.persons.length > 0) {
    const lines = side.persons.map((p) => p.name);
    if (side.unresolved?.kind === 'pending_member') lines.push('partner to be confirmed');
    return lines;
  }
  const u = side.unresolved ?? { kind: 'undetermined' as const };
  switch (u.kind) {
    case 'bye':
      return ['Bye'];
    case 'winner_of':
      return [`Winner of ${u.reference}`];
    case 'loser_of':
      return [`Loser of ${u.reference}`];
    case 'withheld':
      return ['Player not published'];
    case 'pending_member':
      return [...u.known.map((p) => p.name), 'partner to be confirmed'];
    case 'undetermined':
      return ['To be decided'];
  }
}

/** A side collapsed to one line — for the compact chip and the board's own
 *  condensed density (match-card §3.2). Joins with `' / '`, never a rebuilt
 *  string parsed back apart later (D15). */
export function formatSideCondensed(side: Side): string {
  return formatSideLines(side).join(' / ');
}

/** The accessible inline phrase for one side, joining persons with "and"
 *  (match-card §3.2, state-and-formatting §6.1) — never a slash. */
export function sideSummaryText(side: Side): string {
  return formatSideLines(side).join(' and ');
}

/** `"{sideA} versus {sideB}"` — the one accessible-name / inline-summary
 *  builder every renderer uses (match-card §3.2). Built from the structured
 *  sides, never from `persons.map(...).join(' / ')`. */
export function sideSummaryPhrase(sideA: Side, sideB: Side): string {
  return `${sideSummaryText(sideA)} versus ${sideSummaryText(sideB)}`;
}

/** True for a side that is a genuine hole with no label at all — should
 *  never happen (every branch above returns at least one line), kept as a
 *  defensive assertion helper for tests. */
export function isEmptySide(side: Side): boolean {
  return formatSideLines(side).length === 0;
}

/** Preserve the canonical operator-entered display name. Names are not
 *  safely parseable from whitespace, so presentation must never silently
 *  reorder them. See this file's module docstring — a pre-joined-string
 *  compatibility shim, not the structured `Side` formatting above. */
export function formatPersonName(name: string): string {
  return name.trim();
}

/** Format every player inside a pre-joined side string, preserving the
 *  joiner ("A / B" or "A & B"). Compatibility shim — see module docstring. */
export function formatSideName(side: string, joiner: ' / ' | ' & ' = ' / '): string {
  return side.split(joiner).map(formatPersonName).join(joiner);
}

/** The players of a pre-joined side, formatted, one entry per player — for
 *  callers that render each on its own line. Compatibility shim — see module
 *  docstring. */
export function sideNameLines(side: string, joiner: ' / ' | ' & ' = ' / '): string[] {
  return side.split(joiner).map(formatPersonName);
}

File: apps/api/src/shared/sides.py
"""The structured-side wire contract — package 10a (v3 consolidated plan).

Mirrors ``entries/entries_site.py::PersonReferenceDTO`` in shape (an ``id`` +
``name`` reference plus an explicit "why is this not a resolved name" union),
but is a SEPARATE module: the operator wire has no publication gate, and
``shared/`` may not import a domain package (``entries``) per the API's
import-linter contracts. If the two shapes ever need to converge, that is a
decision for whichever package owns publication next — not an import.

``MatchSideDTO`` is the discriminated model the match-card contract
(docs/reference/contracts/match-card.md §2.1) calls ``Side``: 0..n resolved
persons plus an optional ``unresolved`` reason, not mutually exclusive with
``persons`` (a doubles side with one confirmed member is
``persons=[A], unresolved=pending_member(known=[A], missing=1)``).

Per-member resolution (package 29, retiring V3-10-1). ``bracket_participants``
stores only the COMPOSITE team name for a doubles participant
(``entries.py::team_name``, e.g. "Ana Silva / Ben Ito"), but it also stores
``member_ids`` — the roster ids of the two humans, written by the entries
seam for a seam-built pair and by ``ParticipantPicker`` for a hand-entered
one. ``roster_display_names`` reads the SAME roster blob (the workspace ``data``
document's ``bracketPlayers`` list) the public tier already resolves pairs
through (``entries_site.py::_bracket_roster_names``), so a bracket doubles
pair now emits TWO ``PersonRefDTO`` entries with no schema change, no
``entry-{uuid}`` decode and no cross-domain import: the resolver lives here,
in the kernel, and both ``bracket`` and ``entries`` may import it. Where a
member id has no roster row the composite name is still the only honest
label and is emitted as one ``PersonRefDTO``, exactly as before.

``pending_member`` (package 29, retiring V3-10-2) is emitted from two
structural signals, never inferred from a name:

* a TEAM participant carrying exactly ONE member id — a pair slot with a
  member missing outright; and
* an ENTRY-BACKED PLAYER participant in a doubles draw — the entries seam
  drops an entrant whose partner invite is not accepted into the draw as a
  singleton for the director to pair by hand (``entries.py::_pair_batch``
  leg 3: "both carry ``partner_accepted_at`` - a nomination is not a pair"),
  so a lone entry-backed person in an MD/WD/XD draw IS a side one player
  short. An imported or hand-added PLAYER row is deliberately excluded: a
  historical importer may legitimately store a whole pair under one PLAYER
  name, and claiming "partner to be confirmed" over it would be a false
  statement about someone else's draw.
"""
from __future__ import annotations

from typing import List, Literal, Optional

from pydantic import BaseModel, Field

UnresolvedKind = Literal[
    "bye",
    "pending_member",
    "winner_of",
    "loser_of",
    "withheld",
    "undetermined",
]


class PersonRefDTO(BaseModel):
    """One resolved (or dead) person reference on the operator wire.

    Unlike the public ``PersonReferenceDTO`` there is no publication gate
    here — the operator always sees the stored name. ``id`` is the
    roster/participant key the side's ``participantKey`` also carries;
    it is optional because a legacy or hand-entered name may have none.
    """

    id: Optional[str] = None
    name: str


class UnresolvedSideDTO(BaseModel):
    """Why a side has no (or an incomplete) resolved person.

    One flat model rather than six ``kind``-tagged classes: the frontend
    switches on ``kind`` exactly as the match-card contract's
    ``UnresolvedSide`` union describes, and Pydantic serializes the unused
    fields as absent (they carry ``None``/empty defaults), so the wire shape
    for each kind matches the union member the contract names.

    ``reference`` (for ``winner_of`` / ``loser_of``) is the RAW machine id of
    the feeder play unit, not a formatted human reference — the console's
    ``matchIdentity.ts`` / ``bracketLabels.ts`` stay the one authority for
    that spelling (state-and-formatting §6.3, D16) and resolve this id to
    "Winner of QF1" themselves, exactly as they already do for
    ``slot.feeder_play_unit_id`` today.
    """

    kind: UnresolvedKind
    known: List[PersonRefDTO] = Field(default_factory=list)
    missing: int = 0
    reference: Optional[str] = None


class MatchSideDTO(BaseModel):
    """One side of a match — the operator-wire twin of the public ``Side``.

    ``persons`` and ``unresolved`` are NOT mutually exclusive (match-card
    contract §2.1): a partially-known doubles side carries both — see
    ``team_side`` and the ``pending_member`` note in the module docstring.
    """

    persons: List[PersonRefDTO] = Field(default_factory=list)
    unresolved: Optional[UnresolvedSideDTO] = None
    seed: Optional[int] = None
    participantKey: Optional[str] = None


def bye_side() -> MatchSideDTO:
    return MatchSideDTO(unresolved=UnresolvedSideDTO(kind="bye"))


def undetermined_side() -> MatchSideDTO:
    return MatchSideDTO(unresolved=UnresolvedSideDTO(kind="undetermined"))


def winner_of_side(play_unit_id: str, *, loser: bool = False) -> MatchSideDTO:
    return MatchSideDTO(
        unresolved=UnresolvedSideDTO(
            kind="loser_of" if loser else "winner_of",
            reference=play_unit_id,
        )
    )


def resolved_side(
    *, id: Optional[str], name: str, seed: Optional[int] = None
) -> MatchSideDTO:
    return MatchSideDTO(
        persons=[PersonRefDTO(id=id, name=name)],
        seed=seed,
        participantKey=id,
    )


def team_side(
    persons: List[PersonRefDTO],
    *,
    seed: Optional[int] = None,
    participant_key: Optional[str] = None,
    missing: int = 0,
) -> MatchSideDTO:
    """A side of one or more resolved persons, optionally one short.

    ``missing > 0`` attaches the ``pending_member`` reason WITHOUT removing
    the known persons — match-card §2.1's "not mutually exclusive" rule, and
    the only shape that lets a renderer print the known name *and* "partner
    to be confirmed" without inventing a second person.
    """
    unresolved = (
        UnresolvedSideDTO(kind="pending_member", known=list(persons), missing=missing)
        if missing > 0
        else None
    )
    return MatchSideDTO(
        persons=list(persons),
        unresolved=unresolved,
        seed=seed,
        participantKey=participant_key,
    )


# ---------------------------------------------------------------------------
# Roster-name resolution and the pair-discipline test.
#
# Both live in the kernel deliberately: ``bracket`` and ``entries`` each need
# them and neither may import the other (the API's per-domain-independence
# import contracts). Neither reads a table — they take the caller's already
# loaded workspace ``data`` document and a plain discipline code.
# ---------------------------------------------------------------------------

#: Draws whose participants are PAIRS: the three canonical doubles codes
#: plus the two junior spellings the meet side uses (``BD``/``GD``).
PAIR_DISCIPLINES = frozenset({"MD", "WD", "XD", "BD", "GD"})


def is_pair_discipline(discipline: Optional[str], event_id: Optional[str] = None) -> bool:
    """True when a draw's participants are pairs rather than individuals.

    Mirrors ``entries_site.py::_event_public_code``'s fallback: bracket event
    ids are operator-owned and a draw's discipline is sometimes only legible
    from the id's trailing code (``T027-MD``). Anything else — a free-text
    discipline, ``GEN`` — is NOT a pair draw, because guessing one would put
    "partner to be confirmed" under a singles player's name.
    """
    if isinstance(discipline, str) and discipline.strip().upper() in PAIR_DISCIPLINES:
        return True
    if isinstance(event_id, str) and event_id:
        return event_id.rsplit("-", 1)[-1].strip().upper() in PAIR_DISCIPLINES
    return False


def roster_display_names(data_blob: Optional[dict]) -> dict:
    """Bracket-roster id → display name, from the workspace ``data`` blob.

    The one sound way to recover two individual names from a pair: the
    participant's ``member_ids`` are roster ids and this is the roster.
    Splitting the composite team label on ``' / '`` corrupts real names and
    invents identity structure (D15) — that round trip is deleted, not
    standardised. Malformed or unnamed rows stay ABSENT rather than leaking
    an id as though it were a name.

    Twin of ``entries_site.py::_bracket_roster_names``, which applies the
    public tier's extra privacy gates on top of the same blob.
    """
    out: dict = {}
    for row in (data_blob or {}).get("bracketPlayers") or []:
        if not isinstance(row, dict):
            continue
        key, name = row.get("id"), row.get("name")
        if isinstance(key, str) and key and isinstance(name, str) and name.strip():
            out[key] = name.strip()
    return out
